packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -2091,31 +2132,65 @@ class _InactiveElements {
 
   static void _deactivateRecursively(Element element) {
     assert(element._lifecycleState == _ElementLifecycle.active);
-    element.deactivate();
+    try {
+      element.deactivate();
+    } catch (_) {
+      element._ensureDeactivated();
+      rethrow;
+    } finally {
+      assert(element._lifecycleState == _ElementLifecycle.inactive);
+      element.visitChildren(_deactivateRecursively);
+      assert(() {
+        element.debugDeactivated();
+        return true;
+      }());
+    }
+  }
+
+  static void _deactivateFailedSubtreeRecursively(Element element) {
+    // Temporarily set the Element's state to active in case the deactivate
+    // implementation has related assertions.
+    element._lifecycleState = _ElementLifecycle.active;
+    try {
+      element.deactivate();
+    } catch (_) {
+      element._ensureDeactivated();
+      // Do not rethrow:
+      // The subtree has already thrown a different error and the framework is
+      // cleaning up on a best-effort basis.
```

Comment:
The "best-effort basis" here may sound a bit scary, as the failing widget subtree may end up with different states with assertions enabled / disabled (and the framework don't rethrow). But I think the goal is to make sure the Elements are in the correct state as far as the framework is concerned when something fails, to reduce the noise in the crash report. Maybe we should only call `_ensureDeactivated` instead so we don't invoke user deactivate callbacks?

Reply:
> widget subtree may end up with different states with assertions enabled / disabled

won't they all become failed at the end?